# 空指针调用成员函数

## 类

如下类：
``` c++
#include <iostream>
#include <memory>

using namespace std;

class A {
public:
	void foo() {
		cout << "A::foo()" << endl;
	};

	void bar() {
		cout << "i = " << i << endl;
	}

	static void foobar() {
		cout << "A::foobar()" << endl;
	}
private:
	int i;
};
```

## 测试一

``` c++
int main()
{
	A* a = nullptr;
	a->foobar();

	return 0;
}
```

输出:
``` bash
A::foobar()
```

说明：foobar是一个静态成员函数，`a->foobar()` 相当于`A::foobar()`

## 测试二

``` c++
int main()
{
	A* a = nullptr;
	a->foo();

	return 0;
}
```

输出:
``` bash
A::foo()
```

说明: foo是一个成员函数，这里并没有产生想象中的崩溃。在c++中调用成员函数，会将this作为第一个参数，也就是`a->foo()`相当于`foo(a)`,所以相当于执行:
``` c++
void foo(A* a){
    cout << "A::foo()" << endl;
}
```
所以不会产生崩溃。

## 测试三
``` c++
int main()
{
	A* a = nullptr;
	a->bar();

	return 0;
}
```
执行会产生崩溃。

说明: 参考测试二的说明,这里相当于执行:
``` c++
void bar(A* a) {
    cout << "i = " << a->i << endl;
}
```
这里a为nullptr, 所以执行`a->i`时会产生崩溃。

## 总结

执行代码如果崩溃在一些莫名其妙的地方时，可以考虑一下是否遇到来测试三的情形。测试二的情形其实可以等同于测试一，因为没有用到成员变量，所以将foo声明成static也是没有问题的。
